Add datepicker to address registration form accessibility

foxbunny opened the pull request:
- Address registration form accessibility by introducing a custom datepicker
- Add `<ind-datepicker>` custom element

Note: The datepicker supports range selection which links two datepickers so that they limit each other's max/min values. This feature is not used in the registration form, and has thus only been tested out of the context of the Indico application. It "should" work in theory, but it should be considered untested. The feature will be revisited/re-tested when it is integrated into a context where it will be used (e.g., management area).

foxbunny commented:
I still have a few small details to check, but feel free to review. I don't expect any big changes in the code.

ThiefMaster commented on indico/web/client/js/custom_elements/ind_datepicker.js:
```diff
@@ -0,0 +1,631 @@
+  return now;
+})();
+const DEFAULT_LOCALE =
+  Intl.DateTimeFormat?.resolvedOptions?.().locale ||
+  document.documentElement.dataset.canonicalLocale ||
+  document.documentElement.lang ||
+  navigator.language;
```
Is it really a good idea to get this from the browser and not from what Indico considers the current language? I think it'd be weird if someone uses e.g. English Indico on a French browser, and suddenly get French-style date formatting.

Especially since the regform date field in Indico has an option to configure the format...

foxbunny replied:
Ah, yeah, I knew I was forgetting something. I need to add the locale of the page as the `data-canonical-locale` to the `<html>` element in the template and remove the rest of the options. Good catch, thanks.

ThiefMaster replied:
I'd get it by calling `moment.locale()` which should already be set on all pages. That way we don't introduce another source of truth for the current locale...

ThiefMaster commented on CHANGES.rst:
```diff
@@ -293,7 +294,7 @@ Internal Changes
   thanks :user:`foxbunny`)
 - Allow plugins to add extra fields to the room booking form (:pr:`6126`, thanks :user:`VojtechPetru`)
-
+- Add ``<ind-datepicker>`` custom element (:pr:`6371`, thanks :user`foxbunny`)
```
```suggestion
- Add ``<ind-datepicker>`` custom element (:pr:`6371`, thanks :user:`foxbunny`)

```

Fixing the missing `:` in the rst user reference, and preserving the whitespace as-is (we always two blank lines at the end of a version's changelog entries).

ThiefMaster commented on indico/web/client/js/react/components/Datepicker.jsx:
```diff
@@ -0,0 +1,78 @@
+// This file is part of Indico.
```
(I'm starting new threads here for anything I find while testing if I cannot map it to any specific location in the code, simply because it's nicer to reply to compared to top-level comments where GitHub unfortunately has no threading...)

The calendar always opens to the bottom, even if there is not enough space because the triggering element is close to the bottom edge of the page:

![image](https://github.com/indico/indico/assets/179599/1dd059c6-69ce-4bb9-bce7-2dfa9421e7bc)

foxbunny replied:
Which browser (and browser version) are you using? There should be a drop shadow behind the calendar dialog.

ThiefMaster replied:
Firefox 127.0b8 🦊

ThiefMaster replied:
`--control-raised-shadow` isn't set

foxbunny replied:
Yup. It was set on one of the other PRs, forgot to re-check after switching back to master.

ThiefMaster replied:
Did you mean `--modal-control-box-shadow`? I had a try and that looks much better!

![image](https://github.com/indico/indico/assets/179599/f9aafc73-e658-413d-b437-72d36a80134b)

ThiefMaster commented on indico/web/client/js/react/components/Datepicker.jsx:
```diff
@@ -0,0 +1,78 @@
+// This file is part of Indico.
```
The lack of a border around the whole thing makes it a bit hard to see, especially on a page with mostly white background.

ThiefMaster replied:
Clicking the trigger icon again closes and immediately reopens the date picker. I'd expect it to either not do anything, or close the date picker but not reopen it.

ThiefMaster replied:
When I type a date in a different format than the one specified in the regform's field settings, it accepts it (nice!) but it keep showing it in the wrong format. There is no feedback whatsoever. Could we automatically convert the text input to the proper format as soon as the input is valid?

AFAICT the flexibility only affects the separator (`/` vs `.`), but still...

ThiefMaster commented on indico/web/client/js/react/components/Datepicker.jsx:
```diff
@@ -0,0 +1,78 @@
+
+      <span className="date-format">
+        <Translate>
+          Date format: <Param name="format" value={format} />
```
I think this should only be used in the regform field. We don't use this anywhere else in Indico (we use the locale's default date format instead), it's very weird IMHO.

Typically the requested date format is shown as a placeholder in the input field.

ThiefMaster commented on indico/modules/events/registration/client/js/form/fields/DateInput.jsx:
```diff
@@ -42,18 +42,14 @@ export function DateInputComponent({
-          enableOutsideDays
+          format={dateFormat}
+          value={dateValue}
+          onChange={handleDateChange}
```
You also need to pass through the `onFocus` and `onBlur` props from the `DateInputComponent`. That way the final-form field will be correctly marked as "touched" and will immediately show validation errors on blur.

ThiefMaster commented on indico/web/client/js/custom_elements/ind_datepicker.js:
```diff
@@ -0,0 +1,631 @@
+
+// Parser
+
+function createDateParser(formatString) {
```
There's no lib which already does this nicely? I'm fine w/ keeping our own parser (maybe move it to a separate file though?), but we'll certainly need some unit tests for this.

foxbunny replied:
Existing parsers are a bit too strict, so I wrote one that will take anything that approximately looks like the format string. 

> but we'll certainly need some unit tests for this

Sure, I'll add some.

ThiefMaster commented on indico/web/client/js/custom_elements/ind_datepicker.js:
```diff
@@ -0,0 +1,631 @@
+  for (const token of tokens) {
+    switch (token.type) {
+      case TOKEN_YEAR:
+        regexSource += '(?<year>\\d{2}|\\d{4})';
```
I would NOT accept `yy` here. Besides that fact that `yy` generally sucks, I don't think we can safely assume that `24.12.20` is `1920-12-24` (and not `2020-12-24`).

I'd prefer to always require `yyyy` for the year. That way people get a validation error and can fix their date.

ThiefMaster commented on indico/web/client/js/custom_elements/ind_datepicker.js:
```diff
@@ -0,0 +1,631 @@
+
+import './ind_datepicker.scss';
+
+const TODAY = (function() {
```
I know it's a somewhat rare edge case, but this will be incorrect when the day changes between loading the page and opening the date picker ;)

foxbunny replied:
True.

ThiefMaster commented on indico/web/client/js/custom_elements/ind_datepicker.js:
```diff
@@ -0,0 +1,631 @@
+
+// Parser
+
+function createDateParser(formatString) {
```
What do you think about **always** allowing people to type a date in ISO format (`yyyy-mm-dd`) in addition to the chosen date format? Personally I think it'd be nice (since there are no other `yyyy-??-??` style formats we allow to select when configuring the regform field), but maybe that's just the nerd in my who wants to be able to use ISO format instead of whatever weird (`mm/dd/yyyy` 🙈) format someone chose...

If this is not super trivial to implement or you think it's strange, feel free to ignore that suggestion though!

foxbunny replied:
> If this is not super trivial to implement or you think it's strange, feel free to ignore that suggestion though!

Hmmm.... Well, I think we could simply try to parse the date using the parser, and if that fails, try to parse using the Date constructor. Consider it an invalid date only when both fail. Something like that. It shouldn't be too much work.

foxbunny replied:
Hah, learned something new today. The `Date` constructor can parse all kinds of human-readable date (amazing!), but ISO date will be treated as UTC, so that's about the only format that we *cannot* use for this case. 😂 Oh well.

ThiefMaster commented:
> The datepicker supports range selection which links two datepickers so that they limit each other's max/min values.

I think for usecases like room booking but also the accommodation field in the regform, we'll definitely need this - and since you're working on the regform, maybe tackling the accommodation field next would be nice. I recommend trying out the current behavior there, because I think it's more than just "two linked pickers" - we show both pickers at once, and a nice feedback of the range you selected / are selecting:

![image](https://github.com/indico/indico/assets/179599/d0beddae-94e6-40f0-98e1-f75614cfcfce)
![image](https://github.com/indico/indico/assets/179599/4c46aadd-a9e8-4777-9dc9-54d356c76288)

In case of the accommodation field there's also some slightly more tricky login regarding the valid dates, because unlike the typical "min/max number of nights" many date pickers use, we actually have a range of valid arrival and departure dates (which can, of course, be implemented by dynamically generating the mix/max dates like we do in the current version of that field).

foxbunny commented:
I've addressed the feedback from this PR as far as the single-date picker goes. @ThiefMaster Would you like me to tackle the accommodations picker in this PR as well, or do it in a separate PR once this is merged? Makes no difference to me.

ThiefMaster commented:
> Would you like me to tackle the accommodations picker in this PR as well, or do it in a separate PR once this is merged? Makes no difference to me.

Let's see/decide once this one is ready to merge. Separate PR is probably fine though... But let's iron out all the remaining problems with the initial date picker first! :)

ThiefMaster commented on indico/modules/events/registration/client/js/form/fields/DateInput.jsx:
```diff
@@ -42,18 +44,16 @@ export function DateInputComponent({
-          verticalSpacing={10}
-          enableOutsideDays
+          format={dateFormat}
+          value={dateValue}
```
FWIW the way I would expect the Datepicker component to work is this:

- The `value` it receives is always in ISO format
- The `value` it sends (via `onChange`) is always in ISO format
- The displayed value, placeholder and input format (when manually entering a date) uses the specified `format` (and I think ideally that `format` would be optional and default to the locale's date format since the regform is the only place where we want to override the format instead of using the locale's default)

foxbunny replied:
Gotcha. What's the best place to obtain the locale-format mapping?

ThiefMaster replied:
I just discussed w/ @tomasr8 and maybe using momentjs here would be the best option (`moment(...).format('L')`) - we still use it in many places so it's not likely to go away any time soon (even though Intl nowadays supports pretty much everything).

The moment docs also indicate that this is always a number-based format, so it'll probably be fine for all locales:

> Month numeral, day of month, year | L | 09/04/1986

ThiefMaster replied:
`moment.localeData().longDateFormat('L')`

foxbunny replied:
Looks like we can get that information from `moment.localeData().longDateFormat('L')`

ThiefMaster replied:
Not sure why this was marked as resolved, I still get this behavior...

ThiefMaster replied:
I think the problem is that currently the value is going directly into the `defaultValue` of the text input field. Maybe it should be passed e.g. as `<ind-date-picker defaultValue={value}>` instead, so it can be converted to the proper format before it's put in the input field?

foxbunny replied:
I resolved this by mistake. Anyway, fixed the issue by reformatting the date prior to using it as the default value.

foxbunny commented on indico/web/client/js/custom_elements/ind_datepicker.js:
```diff
@@ -0,0 +1,557 @@
+
+const DEFAULT_LOCALE =
+  document.documentElement.dataset.canonicalLocale ||
+  document.documentElement.lang ||
+  navigator.language;
```
It's just for completeness, but usually it shouldn't be needed.

foxbunny commented on indico/util/i18n.py:
```diff
@@ -178,6 +178,11 @@ def weekday(self, daynum, short=True):
         return self.days['format']['abbreviated' if short else 'wide'][daynum]
 
+    @property
+    def html_locale(self):
```
There are a few reasons:

1. The way locale is set for moment is a bit hackish. Rather than basing everything on moment, I would base moment on the newly added attribute instead.
2. Creating a hard-dependency on a library feature will prevent you from removing the library in the long run if, for some reason, the library becomes unusable or platform features supersede it. So I wouldn't recommend it. It also decreases the flexibility of the custom elements, which are currently relying only on platform features as a design choice, to ensure longevity of the code.

Let me know if you still insist on using moment.locale() and I'll update the code.

ThiefMaster replied:
> Rather than basing everything on moment, I would base moment on the newly added attribute instead.

I'd be fine with that, but I'm not sure if it'll work straight away. The moment locales are generated based on what moment contains (see generate_moment_locales.py).

Anyway, maybe best to focus on the other things and leave this one for later?

foxbunny replied:
`session.moment_lang` cannot be used for this because it's missing the territory suffix afaics.

ThiefMaster replied:
ah indeed, because moment uses just the bare language name for the "main" version of a language (e.g. `fr` for french, and `fr-ca` for canadian french)...

foxbunny commented on indico/web/client/js/custom_elements/ind_datepicker.js:
```diff
@@ -0,0 +1,557 @@
+}
+
+function toDateString(date) {
+  // We do not use ISO date format because it will be interpreted as UTC
+  return date?.toLocaleDateString('en-US');
```
> Any way we can avoid putting a non-ISO format in the value of this field, and work around the "will be interpreted as UTC" problem?

These values are used internally. Since the `Date()` constructor parses that format consistently irrespective of the browser's language settings, and it doesn't suffer form the UTC issue, I don't think we should complicate things just for aesthetics.

> It's usually safe to assume that what you select is in the event's timezone since that typically makes sense, but this is not something we can (from a technical point of view) assume / rely on.

The datepicker code is should be time zone agnostic, if that's your concern?

ThiefMaster replied:
That's fine then (I was indeed hoping for it to be TZ agnostic).

So the `.value` of the `<ind-datepicker>` should be considered internal. I was looking at it like a form input field, but maybe I shouldn't do that (or maybe a different name may be better, since I could imagine someone else making the same mistake). If it's internal then I'd say it's fine as long as we do not end up with cases where `mm/dd/yyyy` vs `dd/mm/yyyy` can cause problems

foxbunny replied:
Alternatively we can probably use the RFC date format which is obtained by calling `Date.prototype.toDateString()`, tho.

foxbunny commented on indico/modules/events/registration/client/js/form/fields/DateInput.jsx:
```diff
@@ -42,18 +44,16 @@ export function DateInputComponent({
-          enableOutsideDays
+          format={dateFormat}
+          value={dateValue}
+          onChange={handleDateChange}
```
I'll first address the `value` bit you mentioned above. Then we'll see if that fixes it.

ThiefMaster commented on indico/web/client/js/react/components/Datepicker.jsx:
```diff
@@ -0,0 +1,80 @@
+export default function Datepicker({format, onChange, value, ...inputProps}) {
+  function handleDateChange(ev) {
+    const {date} = ev.target.closest('ind-date-picker');
+    onChange(date?.toLocaleDateString('en-CA'));
```
This may deserve a comment since I guess you use en-CA because they us `YYYY-MM-DD` for dates, but someone not knowing this will probably wonder why we use en-CA here

ThiefMaster commented on indico/web/client/js/react/components/Datepicker.jsx:
```diff
@@ -0,0 +1,83 @@
+
+import 'indico/custom_elements/ind_datepicker';
+
+export default function Datepicker({format, onChange, value, ...inputProps}) {
```
I'd prefer calling it `DatePicker` instead of `Datepicker` for the sake of consistency with existing capitalization of `DatePicker` in the codebase.

ThiefMaster commented on indico/web/client/js/react/components/DatePicker.jsx:
```diff
@@ -0,0 +1,90 @@
+export default function DatePicker({
+  onChange,
+  value,
+  format = moment.localeData().longDateFormat('L'),
```
Does JS evaluate this at calling time or at import time (like Python does)? 

If it's the latter this may be problematic in the future if we end up using this datepicker in room booking, where the moment locale gets set async (since user info is loaded via an API call there).

In that case I'd ather leave it `undefined` here and do something like `format ??= moment.localeData().longDateFormat('L')` in the function body to evaluate it only when this gets actually called.

foxbunny replied:
> In that case I'd ather leave it undefined here and do something like format ??= moment.localeData().longDateFormat('L') in the function body to evaluate it only when this gets actually called.

Default parameters are evaluated only if arguments are not passed, so it would effectively be the same thing. Pretty cool, eh?

ThiefMaster commented on indico/web/client/js/react/components/DatePicker.jsx:
```diff
@@ -0,0 +1,90 @@
+
+            <div className="month-year">
+              <select />
+              <input type="number" min="1900" max="2100" required />
```
I noticed that you can use the calendar to exceed this. For now I think this can be safely ignored, but when adding limitations on the earliest/latest date that can be selected, it should probably be taken into account (and maybe dynamically update the min/max of the year and month pickers here).

foxbunny replied:
Hm, yeah. I'll give it some thought tomorrow.

ThiefMaster replied:
I think the other open threads are higher priority, especially since the base regform field does not need limiting the min/max range anyway (even though the UN "variant" of it, which IIRC is used for birth dates / passport validity dates, might require this). I just left the comment since it's easy to forget such things later ;)

foxbunny replied:
I've decided to just remove the constraints for now. When I start working on the dual-field date range picker, I'll address the min-max constraint anyway. So for now I think this is good to go.

ThiefMaster commented on indico/web/client/js/react/components/DatePicker.jsx:
```diff
@@ -0,0 +1,90 @@
+  return (
+    <ind-date-picker>
+      <input type="text" onChange={handleDateChange} defaultValue={value} {...inputProps} />
+      <button type="button">
```
In case of inputProps.disabled, this needs to be disabled as well.

ThiefMaster commented on indico/web/client/js/react/components/DatePicker.jsx:
```diff
@@ -0,0 +1,99 @@
+
+      <span className="date-format">
+        <Translate>
+          Date format: <Param name="format" value={format} />
```
What do you think about showing this as a `placeholder` in the input field? I think maybe this whole date-format block below could go away altogether then.

foxbunny replied:
Placeholder isn't used by screen readers, so the date format block must be there. That's why I omitted the placeholder.

ThiefMaster replied:
Isn't there a way to hide it from regular users but still expose it to screen readers etc? Maybe using negative margin or whatever hack is often used for things like those skip links at the beginning of the page?

I think especially in places where we use the locale default and thus don't want to show the date format at all, the placeholder would look nicer.

Actually it already looks a bit weird in the registration form when there's also a field description:
![image](https://github.com/indico/indico/assets/179599/e33c2176-e538-4c8e-88bd-f2a042222a2d)

foxbunny replied:
I will make that change. However, I will also note that from the UX perspective, placeholder is generally an anti-pattern. Not a great way of conveying anything. Suppose you start typing 2012 and you get distracted and you get back to the field. You would no longer know what the format should be without removing the input. Not great. If the information matters at all, it should be present outside the field. If it doesn't matter, it should not be present at all.

ThiefMaster replied:
`const formatDescription = this.querySelector('.date-format');` now fails because styleName applies CSS module name mangling

foxbunny replied:
Fixed

ThiefMaster commented on CHANGES.rst:
```diff
@@ -293,7 +294,7 @@ Internal Changes
   thanks :user:`foxbunny`)
 - Allow plugins to add extra fields to the room booking form (:pr:`6126`, thanks :user:`VojtechPetru`)
-
+- Add ``<ind-date-picker>`` custom element (:pr:`6371`, thanks :user:`foxbunny`)
```
```suggestion
- Add ``<ind-date-picker>`` custom element (:pr:`6371`, thanks :user:`foxbunny`)

```

otherwise i'll forget fixing this inconsistency before merging the PR ;)

ThiefMaster commented on indico/web/client/js/custom_elements/ind_date_picker.scss:
```diff
@@ -0,0 +1,165 @@
+  }
+}
+
+ind-calendar {
```
While it's usually me who complains about things being too big and having too much margin/padding, this time I did not have that feeling most of the guys in the team find the date picker a bit too big:

- "Am I the only one who finds the font size/padding of the new datepicker way too big?"
- "This is massive"
- "it looks kinda huge"

Is there anything to tweak w/o reducing a11y? I think on mobile it has the right size (even though personally I usually prefer an `input[type=date]` on mobile, but that's off-topic here) since you can easily tap calendar cells there, but the calendar dialog is indeed quite large.

cc @tomasr8 @GovernmentPlates @duartegalvao

foxbunny replied:
The main issue with the size is actually that styling of the inputs propagates all the way down to the inputs used in the month-year controls. Basically semantic UI applies the input styling too broadly to anything that is inside `.ui.field` or whatever the class is. It can be made smaller if I get rid of the excessive paddings in the select list and text input, but that means more code whose only purpose is to deal with bad design of the Semantic UI. In these cases I generally prefer to err on the side of not having (too many) unnecessary overrides.

foxbunny replied:
Of course, I'm not strictly opposed to adding such overrides if that's what you prefer.

GovernmentPlates replied:
Just thought I would chime in with my own two pence: I'm all for a11y, but as we have already said before, this widget is simply way too big in its current form (not to mention, the size makes this look out-of-place when comparing this to other picker elements we already have), and I personally would rather see this polished/fixed (i.e. reducing the visual size of this) before pushing this up into master.

I know SUI can be a pain in the neck sometimes (this is a recurring theme for all of us here :P - not to mention, other teams at CERN which use React SUI), but please consider fixing this, even if it means doing some annoying overrides.

foxbunny replied:
Sure, no problem.

There appears to be some misunderstanding here, though, so let me try to correct that first.

I agree that the widget is too large. I am also not opposed to fixing that. My point was simply that this adds more technical debt, and from what I see, you're already swimming in it. I just wanted confirmation that you're ok with adding more of it.

ThiefMaster replied:
I think we can't really avoid it... :/
So yeah, we're OK with it

foxbunny replied:
I think it's now reasonably sized. I can make it 12.5% smaller if you'd like, but that would negatively impact the click target size, so I wouldn't recommend that.

I thought a bit about how we can make overrides painfully clear, and decided to implement a few lightweight mixins that you can later hunt down if decide to address debt.

I've also run into this bit of CSS in `_input.scss`:

```css
  &:-moz-focusring {
    color: transparent !important;
    text-shadow: 0 0 0 $black !important;
  }
```

This generally breaks select lists when they're styled normally (outside of SUI). You might want to consider removing it, and apply proper styling for whatever it's trying to "fix".

ThiefMaster replied:
Looks pretty nice now IMHO. Just wondering: Can we easily make this a bit responsive to keep it large on small mobile screens? While it's much better now on desktops, on my phone the large size made it much easier to select a date.

foxbunny replied:
Let me see what I can do.

foxbunny replied:
Ok, so I'd leave this for some other time. Since Indico doesn't scale correctly on mobile devices, it would require some device detection, which just adds flakiness to the design. I don't think we should pursue that solution.

ThiefMaster replied:
Fine for me, it's already better than the old one anyway (tested w/ the accommodation field one).

foxbunny commented on indico/web/client/js/custom_elements/ind_date_picker.scss:
```diff
@@ -0,0 +1,169 @@
+    color: #ddd;
+  }
+
+  .date-format {
+    @extend %visually-hidden;
+  }
```
Crap, got caught by slow build again... Fixed.

foxbunny replied:
I was on the wrong branch when I pushed. Pushed the changes now.

ThiefMaster commented:
I guess this PR is no longer draft considering it seems pretty complete and good to go?

ThiefMaster commented on indico/modules/events/registration/client/js/form/fields/DateInput.jsx:
```diff
@@ -12,7 +12,7 @@ import React from 'react';
 import {Form, Input} from 'semantic-ui-react';
 
-import {SingleDatePicker} from 'indico/react/components';
+import DatePicker from 'indico/react/components/DatePicker';
```
I would add an export to `indico/web/client/js/react/components/index.js` like we do for all the other components and then keep this as:

```
import {DatePicker} from 'indico/react/components';
```

foxbunny replied:
Done

ThiefMaster replied:
Check my last comment again ;) (I think you misunderstood it)

foxbunny replied:
Ahh, ok.

foxbunny replied:
Done 2.0

ThiefMaster commented on indico/web/client/js/react/components/DatePicker.jsx:
```diff
@@ -0,0 +1,100 @@
+            </button>
+
+            <div className="month-year">
+              <select />
```
Looks fine, thanks!

ThiefMaster commented:
Now that this is merged, if you could look into the feature needed for the accommodation field that would be great! Because I think those features might already make it possible to replace the date pickers in other places (e.g. Room Booking) with your new one. And once we can get rid of react-dates for good, we unblock quite a few big JS library upgrades (including using a more recent React version).

Another interesting part to look into may be the time picker. It's used in the regform when you have a date field and add time format in its config, but we have many other places where it's used as well - and while it works OK-ish most of the time, its usability is not really great, and probably its accessibility is pretty poor as well. So I'd love to eventually replace it with something nicer.

tomasr8 commented:
> Another interesting part to look into may be the time picker. It's used in the regform when you have a date field and add time format in its config, but we have many other places where it's used as well - and while it works OK-ish most of the time, its usability is not really great, and probably its accessibility is pretty poor as well. So I'd love to eventually replace it with something nicer.

I think the idea is to use the new [combobox](https://github.com/indico/indico/pull/6310) to create a more accessible time picker

ThiefMaster commented:
Let's use a separate thread for this discussion, I think it's weird to have a possibly extended discussion under this closed PR ;) --> https://github.com/indico/indico/issues/6387

ThiefMaster commented:
When testing the new datepicker elsewhere, I noticed that the styling of its disabled version is quite different from that of SUI elements. Would it be possible to make the gray-out more noticeable like in the SUI elements?

![image](https://github.com/indico/indico/assets/179599/0271f9b6-0bcf-4eb7-8e0c-5349770efe32)

This is mostly used while a form is being submitted, so I *think*/hope strong contrast is less important there...

foxbunny commented:
Can you give me the StR for that scenario, and also maybe push a branch with the integration that you're doing?

ThiefMaster commented:
What does StR mean in this context? This particular screenshot was just me testing after I noticed that it looked differently than the other fields while submitting (=disabled).

ThiefMaster commented:
Actually, I just realized that this particular screenshot may still get some regform CSS (this test was in the "CERN site access" module of one of our custom plugins where I tried changing the birth date field to your new date picker). In other places it still looks slightly different but way better overall:

![image](https://github.com/indico/indico/assets/179599/c9c834ab-53b4-4c0f-9736-97e0be32cf31)

For this particular screenshot I'll add a commit on top of #6394 to test it easily. Replicating the previous screenshot's example w/o having to use that other plugin (which is a bit cumbersome to setup) takes a bit more time. I'll look into it later or tomorrow.

foxbunny commented:
FYI: date picker does not explicitly style its input at all. It takes the appearance of the default input style in the context. Or it should anyway. If the context only uses SUI inputs and doesn't provide any input styling or provides input styling that doesn't match SUI, the datepicker input style will not match SUI.

Also FYI: contrast guidelines does not apply to disabled anything.

foxbunny commented:
> What does StR mean in this context? 

How do I reach the screen you are testing on?

foxbunny commented:
So the immediate goal is to match the disabled state, yes? Also, do you need this ASAP, or is it ok if I first finish the combobox?

ThiefMaster commented:
not urgent at all, it's very minor in fact. And it seems like we're setting disabled in the regform in some weird way so I'll have a closer look there in case that's what's causing this problem altogether (AFAIK in SUI usually the surrounding div.field has the `disabled` class and then the input inside as well, and when only one of them is set then the styling may be off)

foxbunny commented:
Ok, I'm probably spend tomorrow and maybe Monday working on the combobox. I need to rework it so it supports disabled options, and also fix the extra field so it actually works and uses the same ComboBox as the main field.

ThiefMaster commented:
In case you didn't see my message on Matrix: A colleague noticed a bug in Safari: When clicking one of the navigation arrows in the dialog to move between months, clicking them just closes the dialog.
